File: FlutterPlugin/ios/Classes/Native/DaroNativeAdViewFactory.swift
import Flutter
import Daro

class DaroNativeAdViewFactory: NSObject, FlutterPlatformViewFactory {
    private let messenger: FlutterBinaryMessenger
    private var factoryRegistry: [String: DaroNativeAdFactory] = [:]

    init(messenger: FlutterBinaryMessenger) {
        self.messenger = messenger
        super.init()
    }

    func registerFactory(_ factory: DaroNativeAdFactory, withId id: String) {
        factoryRegistry[id] = factory
    }

    func getFactory(withId id: String) -> DaroNativeAdFactory? {
        return factoryRegistry[id]
    }

    func create(
        withFrame frame: CGRect,
        viewIdentifier viewId: Int64,
        arguments args: Any?
    ) -> FlutterPlatformView {
        guard let params = args as? [String: Any],
              let adId = params["adId"] as? Int,
              let platformView = DaroAdInstanceManager.shared.get(forId: adId) else {
            fatalError("Native ad not found in cache. Call load() before mounting widget. Args: \(String(describing: args))")
        }

        return platformView
    }

    func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
        return FlutterStandardMessageCodec.sharedInstance()
    }
}

File: FlutterPlugin/android/src/main/kotlin/so/daro/flutter/native/DaroNativeAdViewFactory.kt
package so.daro.flutter.native

import android.content.Context
import android.view.View
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory
import so.daro.flutter.DaroAdInstanceManager

class DaroNativeAdViewFactory(
    private val context: Context,
    private val messenger: BinaryMessenger
) : PlatformViewFactory(StandardMessageCodec.INSTANCE) {

    private val factoryRegistry = mutableMapOf<String, DaroNativeAdFactory>()

    fun registerFactory(factory: DaroNativeAdFactory, factoryId: String) {
        factoryRegistry[factoryId] = factory
    }

    fun getFactory(factoryId: String): DaroNativeAdFactory? {
        return factoryRegistry[factoryId]
    }

    override fun create(
        context: Context,
        viewId: Int,
        args: Any?
    ): PlatformView {
        val params = args as? Map<*, *> ?: emptyMap<String, Any>()
        val adId = params["adId"] as? Int

        if (adId != null) {
            DaroAdInstanceManager.get(adId)?.let { return it }
        }

        return object : PlatformView {
            override fun getView(): View = View(context)
            override fun dispose() {}
        }
    }
}
